File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> players, int m, int k) {
    int answer = 0;
    queue<int> q; 
    
    for(int t =0; t<24; t++){
        while(!q.empty() && q.front() <= t){ // 큐에 담긴 서버 종료 시각이 종료 시각을 넘었다면
            q.pop();
        }
        
        int required = players[t]/m;
        
        while(required > q.size()){
            int expansion = required - q.size(); // 추가로 필요한 서버 개수
            answer += expansion;
            
            for(int i =0; i<expansion; i++){
                q.push(t+k); // 서버 종료 시각 넣어서 추가
            }
        }
        
    }
    
    
    return answer;
}

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.*;

class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        Queue<Integer> q = new LinkedList<>();
        for(int t =0; t<24; ++t){
            while(!q.isEmpty() && q.peek() <=t){
                q.poll();
            }
            
            int required = players[t]/m;
            while(required > q.size()){
               int expansion = required - q.size();
                answer += expansion;
                
                for(int i =0; i<expansion; ++i){
                    q.add(t+k);
                }
            }
        }
        return answer;
    }
}